src/v/datalake/values_protobuf.cc

```diff
@@ -0,0 +1,333 @@
+#include "datalake/protobuf_utils.h"
+#include "iceberg/values.h"
+#include "ssx/future-util.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <fmt/core.h>
+
+namespace datalake {
+namespace pb = google::protobuf;
+namespace parsed = serde::pb::parsed;
+namespace {
+
+value_conversion_exception
+type_conversion_error(const pb::FieldDescriptor& fd) {
+    return value_conversion_exception(fmt::format(
+      "Protocol buffers type '{}' conversion is not supported",
+      fd.type_name()));
+}
+
+template<typename BaseT, typename IcebergT, typename DefaultF>
+auto convert(
+  std::optional<parsed::message::field> f,
+  const pb::FieldDescriptor& fd,
+  DefaultF get_default) {
+    if (f.has_value()) {
+        return IcebergT(std::get<BaseT>(std::move(f.value())));
+    }
+    if constexpr (std::is_same_v<IcebergT, iceberg::string_value>) {
+        return IcebergT(iobuf::from(std::invoke(get_default, &fd)));
+    } else {
+        return IcebergT(std::invoke(get_default, &fd));
+    }
+}
+
+// converts a struct to an iceberg value
+ss::future<value_outcome> message_to_value(
+  std::unique_ptr<parsed::message> message,
+  const pb::Descriptor& descriptor,
+  proto_descriptors_stack& stack);
+
+// converts a struct field to an iceberg value, it supports maps and repeated
+// fields
+ss::future<value_outcome> message_field_to_value(
+  std::optional<parsed::message::field> field,
+  const pb::FieldDescriptor& field_descriptor,
+  proto_descriptors_stack& stack);
+
+// converts a single element to iceberg value
+ss::future<value_outcome> single_field_to_value(
+  std::optional<parsed::message::field> field,
+  const pb::FieldDescriptor& field_descriptor,
+  proto_descriptors_stack& stack);
+
+template<typename SourceT>
+ss::future<value_outcome> convert_repeated(
+  chunked_vector<SourceT> list,
+  const pb::FieldDescriptor& field_descriptor,
+  proto_descriptors_stack& stack) {
+    auto ret = std::make_unique<iceberg::list_value>();
+    ret->elements.reserve(list.size());
+
+    for (auto& element : list) {
+        auto result = co_await single_field_to_value(
+          std::move(element), field_descriptor, stack);
+        if (result.has_error()) {
+            co_return result.error();
+        }
+        ret->elements.push_back(std::move(result.value()));
+    };
+    co_return value_outcome{std::move(ret)};
+}
+
+ss::future<value_outcome> convert_repeated(
+  chunked_vector<bool> list,
+  const pb::FieldDescriptor& field_descriptor,
+  proto_descriptors_stack& stack) {
+    auto ret = std::make_unique<iceberg::list_value>();
+    ret->elements.reserve(list.size());
+
+    for (auto element : list) {
+        auto result = co_await single_field_to_value(
+          element, field_descriptor, stack);
+        if (result.has_error()) {
+            co_return result.error();
+        }
+        ret->elements.push_back(std::move(result.value()));
+    };
+    co_return value_outcome{std::move(ret)};
+}
```

Comment:
Why do we need this overload? Is it because of the reference stuff doesn't work? Annoying

Reply:
exactly it seems that `bool` is special treated: 
https://godbolt.org/z/nrT5Gv6GG

Reply:
Maybe we can use `chunked_vector<T>::reference` to consolidate the code: https://github.com/redpanda-data/redpanda/blob/4c2c24c468048905a00d4f52bc546e97ad88bcbf/src/v/container/fragmented_vector.h#L75